core/src/util/builders.cpp

```diff
@@ -384,32 +396,50 @@ void buildPolyLineSegment(const Line& _line, PolyLineBuilder& _ctx) {
     addPolyLineVertex(coordNext, normNext, {1.f, 1.f}, _ctx); // right corner
     addPolyLineVertex(coordNext, -normNext, {0.f, 1.f}, _ctx); // left corner
     indexPairs(1, _ctx.numVertices, _ctx.indices);
-    addCap(coordNext, normNext, cornersOnCap , false, _ctx);
+    if (endCap) {
+        addCap(coordNext, normNext, cornersOnCap, false, _ctx);
+    }
 
 }
 
 void Builders::buildPolyLine(const Line& _line, PolyLineBuilder& _ctx) {
 
+    size_t lineSize = _line.size();
+
     if (_ctx.keepTileEdges) {
 
-        buildPolyLineSegment(_line, _ctx);
+        buildPolyLineSegment(_line, _ctx, 0, lineSize);
 
     } else {
 
         int cut = 0;
+        int firstCutEnd = 0;
 
-        for (size_t i = 0; i < _line.size() - 1; i++) {
+        // Determine cuts
+        for (size_t i = 0; i < lineSize - 1; i++) {
             const glm::vec3& coordCurr = _line[i];
             const glm::vec3& coordNext = _line[i+1];
             if (isOutsideTile(coordCurr, coordNext)) {
-                Line line = Line(&_line[cut], &_line[i+1]);
-                buildPolyLineSegment(line, _ctx);
+                if (cut == 0) {
+                    firstCutEnd = i + 1;
+                }
+                buildPolyLineSegment(_line, _ctx, cut, i + 1);
                 cut = i + 1;
             }
         }
 
-        Line line = Line(&_line[cut], &_line[_line.size()]);
-        buildPolyLineSegment(line, _ctx);
+        if (_ctx.closedPolygon) {
+            if (cut == 0) {
+                // no tile edge cuts!
+                // loop and close the polygon with no endcaps
+                buildPolyLineSegment(_line, _ctx, 0, lineSize+2, false);
```

Comment:
Discussed in IRL: This extra line segment will not be of huge waste, and GL pipeline will be culling this anyhow!